src/Compilers/Core/Portable/RealParser.cs

```diff
@@ -358,14 +342,24 @@ public static DecimalFloatingPointString FromSource(string source)
                     int firstExponent = i;
                     int lastExponent = i;
                     while (i < source.Length && source[i] >= '0' && source[i] <= '9') lastExponent = ++i;
-                    int exponentMagnitude = int.Parse(source.Substring(firstExponent, lastExponent - firstExponent));
-                    if (exponentSign == '-')
+
+                    int exponentMagnitude = 0;
+
+                    if (int.TryParse(source.Substring(firstExponent, lastExponent - firstExponent), out exponentMagnitude) &&
+                        exponentMagnitude <= MAX_EXP_QUAD)
                     {
-                        exponent -= exponentMagnitude;
+                        if (exponentSign == '-')
+                        {
+                            exponent -= exponentMagnitude;
+                        }
+                        else
+                        {
+                            exponent += exponentMagnitude;
```

Comment:
@leppie not so. Given your
```cs
                   const int MAX_EXP_QUAD = (1 << 14); // IEEE quad, without sign
```
That is 16 thousand, not 2.1 billion. I don't have a problem with that being the limit on the allowable total exponent for converted reals. I have a problem with a number 16,000 digits long *along with* an exponent of -16,400 turning into an error instead of a correctly-parsed floating-point number.